Methods for ActiveSupport::Duration parsing from ISO 8601 and output to it

Envek opened the pull request:
This PR add methods to `ActiveSupport::Duration` to allow present it in [ISO 8601 Duration](http://en.wikipedia.org/wiki/ISO_8601#Durations) format and instantiate from it.

[ISO 8601 Duration](http://en.wikipedia.org/wiki/ISO_8601#Durations) format is standartized way to represent duration for interchange, it's already recognized by some database engines (e.g. PostgreSQL) and client side libraries (e.g. [plugin for Moment.js durations](https://github.com/dordille/moment-isoduration)). So, I think it should be part of `ActiveSupport::Duration`.

Some parts of code and tests are taken from [ISO8601 gem](https://github.com/arnau/ISO8601) by Arnau Siches (@arnau) and contributors. Many thanks to them.

This PR is required for PostgreSQL `interval` datatype support (for converting it from and to `ActiveSupport::Duration`) as I've proposed in Google Group [here](https://groups.google.com/forum/#!topic/rubyonrails-core/zcscu1p4ucA). See #16919. Because of that my parsing allows individual datetime parts to be negative (see [PostgreSQL interval output](http://www.postgresql.org/docs/9.4/static/datatype-datetime.html#DATATYPE-INTERVAL-OUTPUT)).

There is no backward incompatible changes as I want it to be included in upcoming 4.2 release, if it still possible (pleeeaasee!).

Envek committed:
Methods for ActiveSupport::Duration parsing from and output to ISO 8601

This is required for PostgreSQL interval datatype support and its convertion from and to ActiveSupport::Duration.
Also it may be useful for others. Some parts of code and tests are taken from ISO8601 gem by Arnau Siches (@arnau).

[Andrey Novikov, Arnau Siches]

Envek commented:
To avoid circular require warnings (absolutely don't understand why are they appearing [such a lot](https://travis-ci.org/rails/rails/jobs/49959946)) included ISO8601 parser into file `activesupport/lib/active_support/duration.rb`. Squashed and rebased on top of current master. Please review one more time.

kaspth commented:
Did you require duration in the parser and vice versa? That sounds like it could cause the circular requires.

Envek commented:
No, I had required parser only from duration (no requires from file with parser at all): [commit that caused a lot of warnings](https://github.com/Envek/rails/commit/929ed6f0840d76f4363032ac69e6e843fa86827d).

kaspth commented:
`Duration` was being autoloaded. It hit `class Duration` in the parser which would trigger another autoload and thus the files would keep trying to load each other.

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
+          raise ParsingError.new("Invalid ISO 8601 duration: #{iso8601duration} (only last part can be fractional)")
+        end
+      end
+
```
:scissors: this line

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
+      end
+
+    end
+
```
:scissors: this too

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601DurationParser
```
The indentation level should match #method_missing above.

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
+              ) # Duration
+            $/x) || raise(ParsingError.new("Invalid ISO 8601 duration: #{iso8601duration}"))
+        sign = match[:sign] == '-' ? -1 : 1
+        @parts = match.names.zip(match.captures).reject{|_k,v| v.nil? }.map do |k, v|
```
I'd extract this to a private method. Also we prefer spaces in blocks `reject { |_, v| v.nil? }`.

Envek replied:
What exactly should go into private method? Only `@parts` construction from regexp match? Or validations too?

kaspth replied:
I'd extract all of those into separate private methods.

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
+        end
+        @parts = ::Hash[parts].slice(:years, :months, :weeks, :days, :hours, :minutes, :seconds)
+        # Validate that is not empty duration or time part is empty if 'T' marker present
+        if parts.empty? || (match[:time].present? && match[:time][1..-1].empty?)
```
You switch back and forth from using parts the instance variable and the reader method. Ditch the `attr_reader` and stick to `@parts`.

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601DurationParser
+      attr_reader :parts
```
After sticking to `@parts` just delete this.

Envek replied:
It's a kind of this parser's API. That parts are being fed to duration in `ActiveSupport::Duration.parse!` method (see line 111)

kaspth replied:
Ah, didn't catch that :+1:

gsamokovarov commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+        (3.years + 3.days).iso8601
+        # => "P3Y3D"
+
+    *Andrey Novikov*
```
I think it'd be fair to put Arnau Siches name here as well. This can also act as a reminder for us to special case it in the contributors app.

kaspth replied:
@Envek you can also add a reference in the commit using [] around your comma separated names then I believe GitHub will attribute you both.

I don't know if the contributors app recognizes this though (cc @fxn).

Envek replied:
I've tried to feed in `git commit`'s `--author` option various combinations like `Andrey Novikov <envek@envek.name>, Arnau Siches <asiches@gmail.com>` or `[Andrey Novikov <envek@envek.name> Arnau Siches <asiches@gmail.com>]`, but no luck: git always strips everything after first email. Git just doesn't support multiple authors.
Pushed commit with author `Andrey Novikov, Arnau Siches <envek@envek.name>`

fxn commented:
@kaspth the contrib app certainly understands a trailing `[...]` with authors separated by commas and other separators. See for example https://github.com/rails/rails/commit/6f8d9bd6da6349d3d179f2e72db5bc7044a8e5c1.

I didn't know GitHub had that feature... you sure it does?

Envek commented:
I've accounted all the notes and force pushed changes in the same commit. I've tried split initializer on separate methods, but dislike current implementations (they are linked too tight with each other).

P.S> in [6f8d9bd](https://github.com/rails/rails/commit/6f8d9bd6da6349d3d179f2e72db5bc7044a8e5c1.patch) only one author specified. Git doesn't support multiple authors (at least with each one's email)

kaspth commented:
@fxn I did remember seeing commits like that. I assumed it was for GitHub and not the contributors app, but it was the other way. Sorry about that!

Envek commented:
Ah, sorry, misunderstood you. Added both authors in `[]` in end of commit message.

fxn commented:
@Envek If the body of the commit message ends with

```
[Andrey Novikov, Arnau Siches]
```

at the very bottom, last line, the contrib app will pick it up automatically and give both of you credit.

See for example https://github.com/rails/rails/commit/84c0f73c8daf50fa98d1c4a0c1bab8708e49d0e4. That one uses "&" but a comma also works.

fxn commented:
@kaspth cool :)

fxn commented:
@Envek looking good!

kaspth commented:
@fxn do you have time for an autoloading question? @Envek was seeing circular require errors: https://github.com/rails/rails/pull/16917#issuecomment-73422799

fxn commented:
Looks like a regular Ruby warning for circular requires (an autoloading error would look [like this](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/dependencies.rb#L492)). I can't tonight but will have a look as soon as I can.

kaspth commented:
@fxn Thanks :)

Envek commented:
Also I've noticed that currently `ActiveSupport::Duration#to_s` (and, thus, `to_json`) method returns just a number of seconds, but it may be more appropriate to return ISO 8601 formatted duration string [as it recommended by Google JSON Style Guide](https://google-styleguide.googlecode.com/svn/trunk/jsoncstyleguide.xml?showone=Time_Duration_Property_Values#Time_Duration_Property_Values) for example. Also there is some support in JS libraries (see [moment.js duration asJson()](http://momentjs.com/docs/#/durations/as-json/) and [moment-duration](https://github.com/luisfarzati/moment-interval#duration-momentdurationstring)).

So, questions:

 1. Should I add support for this in this PR (while it is not merged)
 2. How I should do that (should I add `to_json` method or `as_json` method, I'm confused with them).

Envek committed:
Extract duration parser into separate file

Envek committed:
Changed duration output as json to ISO 8601 duration format

Envek commented:
I have rebased changes on top of current master and also
 1. Extracted parser to separate file and require it only when it first time used. Circular require warnings are gone.  [Commit](https://github.com/Envek/rails/commit/d050b26da5043c161a93b01e4096dd4f6104e550). **Question**: is it fine to require inside method? I've seen such requires somewhere in rails codebase.
 2. Changed `as_json` to return ISO8601 formatted string. This is backward incompatible. [Commit](https://github.com/rails/rails/commit/9f12c3768639e5db8eda93286a2e0ab668c51350). **Question**: Should I keep or delete this change? Or may be also add a note to upgrade guide or somewhere?

I placed these changes in separate commits for review convenience.

kaspth commented on activesupport/lib/active_support/duration/iso8601_duration_parser.rb:
```diff
@@ -0,0 +1,61 @@
+# See http://en.wikipedia.org/wiki/ISO_8601#Durations
+# Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+# This parser isn't so strict and allows negative parts to be present in pattern.
+class ActiveSupport::Duration::ISO8601DurationParser
```
Let's do it this way:

```ruby
module ActiveSupport
  module Duration
    class ISO8601Parser
    end
  end
end
```

I'd also like the parser to be renamed as `ISO8601Parser`. It's already scoped to the `Duration` module, so I don't think we need to repeat that.

kaspth commented on activesupport/lib/active_support/duration/iso8601_duration_parser.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  private
+
```
:scissors: this line and indent under private

kaspth commented on activesupport/lib/active_support/duration/iso8601_duration_parser.rb:
```diff
@@ -0,0 +1,61 @@
+      [ k.to_sym, sign * value ]
+    end
+    @parts = ::Hash[@parts].slice(:years, :months, :weeks, :days, :hours, :minutes, :seconds)
+  end
```
Won't something like this work?

```ruby
def construct_parts
  sign = @match[:sign] == '-' ? -1 : 1

  @parts = @match.names.zip(@match.captures).each_with_object({}) do |key, capture, parts|
    if capture
      part = capture =~ /\d+[\.,]\d+/ ? capture.sub(',', '.').to_f : capture.to_i

      parts[key.to_sym] = sign * part
    end
  end
end
```

kaspth commented on activesupport/lib/active_support/duration/iso8601_duration_parser.rb:
```diff
@@ -0,0 +1,61 @@
+  def construct_parts
+    sign = @match[:sign] == '-' ? -1 : 1
+    @parts = @match.names.zip(@match.captures).reject { |_k,v| v.nil? }.map do |k, v|
+      value = /\d+[\.,]\d+/ =~ v ? v.sub(',', '.').to_f : v.to_i
```
Is `sub` used to satisfy the spec? Otherwise I think we should use `I18n` here.

Envek replied:
This is to satisfy the spec. As referenced by wikipedia in [ISO 8601](http://en.wikipedia.org/wiki/ISO_8601#cite_note-15):

> ISO 8601:2004(E), ISO, 2004-12-01, 4.2.2.4 ... the decimal fraction shall be divided from the integer part by the decimal sign specified in ISO 31-0, i.e. the comma [,] or full stop [.]. Of these, the comma is the preferred sign.

There is no place for `I18n` because format is not locale dependent

kaspth replied:
:+1:

kaspth commented on activesupport/lib/active_support/duration/iso8601_duration_parser.rb:
```diff
@@ -0,0 +1,61 @@
+  # Checks for various semantic errors as stated in ISO 8601 standard
+  def validate_parts!
+    # Validate that is not empty duration or time part is not empty if 'T' marker present
+    if @parts.empty? || (@match[:time].present? && @match[:time][1..-1].empty?)
```
I think we should split these out into two raises.

The "empty duration" doesn't make sense to me.

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -89,21 +89,82 @@ def ago(time = ::Time.current)
+
+    # Build ISO 8601 Duration string for this duration.
+    # The +precision+ parameter can be used to limit seconds' precision of duration.
+    def iso8601(precision=nil)
```
Let's push this logic to the parser and turn precision into a keyword argument here.

Envek replied:
I don't think that this is a _parser_'s job to format existing durations.
But keyword argument is nice — turned.

kaspth replied:
That's true to a degree. The parser is an iso8601 parser, so it would make sense for it to handle the iso8601 parts.

I mean Duration isn't meant to handle iso8601 that's why you made a separate class :smile:

There's also plenty of precedent to have a Serializer class handle both serialization and deserialization. For instance Active Support's MessageVerifier and MessageEncryptor, contrary to their names, verify & sign and encrypt & decrypt, respectively.

Envek replied:
Well, I just [have told to extract this](https://github.com/rails/rails/pull/16917#discussion_r17522955) to separate class (as parsing takes many kines of code due to various validations and so on). For me it looks natural for class to be able to (de)serialize its objects from/to standard format.
I will think about that and probably I will create a new class, something like `ActiveSupport::Duration::ISO8601Serializer` to handle only export of durations to strings in standartized manner :-) (currently, `ISO8601Parser` does not know anything about `Duration`, it just creates compatible data structure to be fed to `Duration`'s initializer)

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -89,21 +89,82 @@ def ago(time = ::Time.current)
+        p[k] += v  unless v.zero?
+        p
+      end
+      parts.default = nil
```
Let's find a better way to write this. Messing with the default makes it feel like a hack.

Envek committed:
Fixed notes from @kaspth (except https://github.com/rails/rails/pull/16917#discussion_r25768661)

Envek commented:
Fixed most of the notes.

kaspth commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -89,21 +89,83 @@ def ago(time = ::Time.current)
+          p[k]  += v
+        end
+        p
+      end
```
I still find this hard to read. Would `transform_values` help us here?

Envek replied:
No. `parts`is not a `Hash`, it is an `Array` of key-value pairs (arrays) with (may be) repetitive keys. Task is summarize values for repetitive keys.

gsamokovarov replied:
Can we use a hash default here? Something like:

```ruby
def normalized_parts
  parts.each_with_object(Hash.new(0)) { |(k, v), p| p[k] += v }
end
```

kaspth replied:
I agree on the `each_with_object`, but I did advise against the default earlier because it was set to nil after the iteration (and therefore felt like a hack to me).

gsamokovarov replied:
I don't know the context well enough, but do we care about resetting the default? Can't we test against zero time part instead of time part presence?

Envek replied:
I've tested against zero in beginning, but have told not to do so there: https://github.com/rails/rails/pull/16917#discussion_r17522966

gsamokovarov replied:
Cool, so maybe an improvement we can make here is just `each_with_object`.

Envek replied:
What about to keep zero as a Hash default and later when constructing a string to check for a key presence?
```ruby
def normalized_parts
  parts.each_with_object(Hash.new(0)) { |(k, v), p| p[k] += v }
end

def iso8601
# ...
output << "#{parts[:years]}Y"   if parts.key?(:years)
# ...
end
```

gsamokovarov replied:
Pinging @jeremy. Are you okay with that?

gsamokovarov commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -89,21 +89,83 @@ def ago(time = ::Time.current)
+      output, sign = 'P', ''
+      parts = normalized_parts
+      # If all parts are negative - let's output negative duration
+      if parts.values.compact.all?{|v| v < 0 }
```
This bit makes me wonder, whether it doesn't belong in the normalization as well. We still have to know the sign then, so maybe it can be given as a return result from the normalization utility.

kaspth replied:
Agree :+1:

> Den 09/03/2015 kl. 11.27 skrev Genadi Samokovarov <notifications@github.com>:
> 
> In activesupport/lib/active_support/duration.rb <https://github.com/rails/rails/pull/16917#discussion_r26025314>:
> 
> > +    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
> > +    # This method allows negative parts to be present in pattern.
> > +    # If invalid string is provided, nil will be returned.
> > +    def self.parse(iso8601duration)
> > +      parse!(iso8601duration)
> > +    rescue ISO8601Parser::ParsingError
> > +      nil
> > +    end
> > +
> > +    # Build ISO 8601 Duration string for this duration.
> > +    # The +precision+ parameter can be used to limit seconds' precision of duration.
> > +    def iso8601(precision: nil)
> > +      output, sign = 'P', ''
> > +      parts = normalized_parts
> > +      # If all parts are negative - let's output negative duration
> > +      if parts.values.compact.all?{|v| v < 0 }
> This bit makes me wonder, whether it doesn't belong in the normalization as well. We still have to know the sign then, so maybe it can be given as a return result from the normalization utility.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/16917/files#r26025314>.
> 

--
Kasper

Envek replied:
Done

Envek committed:
Extracted ISO8601 serialization to specialized class

Envek commented:
Extracted serialization to separate class.

Envek commented:
Can anyone take a look at this again? Are there any issues remaining? I would very much like to finish it and switch to #16919

pixeltrix commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -89,19 +89,52 @@ def ago(time = ::Time.current)
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601Parser::ParsingError+.
+    def self.parse!(iso8601duration)
+      require_relative 'duration/iso8601_parser' unless defined?(ISO8601Parser)
```
I'm sure there's no need to have these requires here - there should be a way to get this to work without them.

Envek replied:
I've not found any way to place them on top of file without triggering **a lot** of circular require warnings. See discussion starting at https://github.com/rails/rails/pull/16917#issuecomment-55709527

pixeltrix replied:
@Envek that implies that there's something wrong with how we've got things setup - not so much with your code. We should investigate what those are rather than committing it with these workarounds.

Envek replied:
@pixeltrix How about `autoload`? (see last commit, that I've added)

Envek committed:
Switch from require_relative in methods to autoload at start of class

Envek commented:
For some reason Travis did not tested last commit, so I rebased whole branch to fresh master and pushed.

@pixeltrix can you review one more time? I've replaced `require`'s inside methods to `autoload` at the beginning of class (month ago). Looks much better, no circular require warnings. See https://github.com/Envek/rails/commit/20b02b35552732c2480d9591cbb7cd1f41f61249

Envek commented:
Please can anyone review again? This PR is required for #16919 which is scheduled for 5.0.0 and I want to continue work on it.

pixeltrix commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601Parser::ParsingError+.
+    def self.parse!(iso8601duration)
+      parts = ISO8601Parser.new(iso8601duration).parts
+      time  = ::Time.now
```
Shouldn't this be `Time.current`?

Envek replied:
@pixeltrix yes, should be `Time.current`. Fixed!

fphilipe replied:
Does it make a difference? The distance of two time in two timezones is still the same.

```ruby
Time.zone = 'US/Pacific'
# Local time zone is UTC
Time.now - Time.current # => 0 (almost)

Envek committed:
Changed Time.now to Time.current

pixeltrix commented on activesupport/lib/active_support/duration/iso8601_parser.rb:
```diff
@@ -0,0 +1,72 @@
+      private
+        def parse_iso_duration!
+          @match = @raw.match(/^
+                      (?<sign>\+|-)?
```
Does the actual ISO 8601 standard allow for a sign character - the wikipedia article doesn't mention them

Envek replied:
At least in the wild there are negative durations possible and they may be useful in math with dates and times:
For example in ActiveSupport: `-1.second` 
For example in PostgreSQL: http://www.postgresql.org/docs/9.4/static/datatype-datetime.html

pixeltrix replied:
And what about individual components being negative, e.g. `P1M-1D` - have you seen any of those in the wild?

Envek replied:
I haven't seen any and don't want to use such scary things, but it's got covered as well. Someone definitely will need it someday.
```
> d = ActiveSupport::Duration.parse('P-1Y-2M3DT-4H-5M-6S')
=> -1 years, -2 months, 3 days, -4 hours, -5 minutes, and -6 seconds
> now = Time.current
=> Tue, 03 Nov 2015 19:54:50 MSK +03:00
> now + d
=> Sat, 06 Sep 2014 15:49:44 MSK +04:00
> now - 1.year - 2.month + 3.days - 4.hours - 5.minutes - 6.seconds
=> Sat, 06 Sep 2014 15:49:44 MSK +04:00
```

arnau replied:
@pixeltrix my take on this is that the sign belongs to the duration not to the component as it is not about expressing complex durations (1M - 1D ~ 29ish days?) but allowing basic arithmetics with durations.  So `2015-11-02 - P1Y = 2014-11-02`.

Envek replied:
This PR is a basis for another PR #16919 to work with interval datatype of PostgreSQL, and Postgres docs provides examples of such scary durations, so we should support them. See http://www.postgresql.org/docs/9.4/static/datatype-datetime.html#DATATYPE-INTERVAL-OUTPUT at most bottom. Postgres usually well known for it's strictness about everything.

arnau replied:
If the goal is to support Postgres current implementation then I agree with supporting anything they support no matter what the ISO says.

That said, I just checked the ISO 8601:2004 spec and I failed to find anything close to this "mixed interval".  Any form of duration with or without recurrence is expressed as a fixed amount not as a computation of its components.

pixeltrix commented:
I'm not a big fan of using a complicated regexp as a parser - I'd rather we'd use `StringScanner` which is designed for this kind of thing. For example here's one that I've made:

``` ruby
require 'strscan'

module ActiveSupport
  class Duration
    class ParseError < ArgumentError; end

    class Parser
      PERIOD_OR_COMMA = /\.|,/
      PERIOD = '.'.freeze
      COMMA = ','.freeze

      SIGN_MARKER = /\A\-|\+|/
      DATE_MARKER = /P/
      TIME_MARKER = /T/
      DATE_COMPONENT = /(\d+(?:[.,]\d+)?)(Y|M|D|W)/
      TIME_COMPONENT = /(\d+(?:[.,]\d+)?)(H|M|S)/

      DATE_TO_PART = { 'Y' => :years, 'M' => :months, 'W' => :weeks, 'D' => :days }
      TIME_TO_PART = { 'H' => :hours, 'M' => :minutes, 'S' => :seconds }

      DATE_COMPONENTS = [:years, :months, :days]
      TIME_COMPONENTS = [:hours, :minutes, :seconds]

      class << self
        def parse!(string)
          new(string).parse!
        end

        def parse(string)
          begin
            parse!(string)
          rescue ParseError => e
            nil
          end
        end
      end

      attr_reader :parts, :scanner
      attr_accessor :mode, :sign

      def initialize(string)
        @scanner = StringScanner.new(string)
        @parts = {}
        @mode = :start
        @sign = 1
      end

      def parse!
        while !finished?
          case mode
          when :start
            if scan(SIGN_MARKER)
              self.sign = matched == '-' ? -1 : 1
              self.mode = :sign
            else
              raise_parse_error
            end

          when :sign
            if @scanner.scan(DATE_MARKER)
              @mode = :date
            else
              raise_parse_error
            end

          when :date
            if scan(TIME_MARKER)
              self.mode = :time
            elsif scan(DATE_COMPONENT)
              parts[DATE_TO_PART[scanner[2]]] = number * sign
            else
              raise_parse_error
            end

          when :time
            if scan(TIME_COMPONENT)
              parts[TIME_TO_PART[scanner[2]]] = number * sign
            else
              raise_parse_error
            end

          end
        end

        valid? ? parts : raise_parse_error
      end

      private

      def finished?
        scanner.eos?
      end

      def matched
        scanner.matched
      end

      def number
        scanner[1] =~ PERIOD_OR_COMMA ? scanner[1].tr(COMMA, PERIOD).to_f : scanner[1].to_i
      end

      def scan(pattern)
        scanner.scan(pattern)
      end

      def string
        scanner.string
      end

      def raise_parse_error
        raise ParseError, "Invalid ISO 8601 duration: #{string}"
      end

      def valid?
        return false if parts.empty? # Empty duration

        # Mixing any of Y, M, D with W is invalid
        if parts.key?(:weeks)
          return false unless (parts.keys & DATE_COMPONENTS).empty?
        end

        # Specifying an empty T part is invalid
        if mode == :time
          return false if (parts.keys & TIME_COMPONENTS).empty?
        end

        return true
      end
    end
  end
end
```

I've not tested it thoroughly so it may need tweaking. Ideally you'd use something like Racc and write a parser grammar but I think in this case it may be overkill.

arnau commented:
@pixeltrix it looks interesting. I might port that approach to https://github.com/arnau/ISO8601 if you don't mind :)

pixeltrix commented:
@arnau feel free to :smile:

Envek commented:
I will now try to implement, visually compare, and benchmark both approaches. Will post results here in a hour or two, I think.

Envek commented:
I took a StringScanner implementation by @pixeltrix , adopted it a bit to make tests pass and pushed to ba7a301

I have benchmarked it with next snipped run with command `cd rails/activesupport && bundle exec irb`

```ruby
require_relative 'lib/active_support/duration'
require 'benchmark/ips'
Benchmark.ips do |x|
  x.report("regex parser") { ActiveSupport::Duration.regex_parse('P1Y2M3DT4H5M6S') }
  x.report("string parser") { ActiveSupport::Duration.str_parse('P1Y2M3DT4H5M6S') }
  x.compare!
end
```

And got such results:

```
Calculating -------------------------------------
        regex parser     1.679k i/100ms
       string parser     1.887k i/100ms
-------------------------------------------------
        regex parser     17.704k (± 5.2%) i/s -     88.987k
       string parser     20.686k (± 4.6%) i/s -    103.785k

Comparison:
       string parser:    20685.8 i/s
        regex parser:    17703.9 i/s - 1.17x slower
```

So switching to StringScanner seems reasonable. Will adopt it further and push to this PR's branch. A hour or two may be. Stay tuned.

Envek committed:
Added messages to asserts inside loops for ease of debugging

Envek committed:
Added test cases for durations with different signs on components

arnau commented:
@Envek nice one!

pixeltrix commented:
@Envek nice - it'd be interesting to see the object count as well - how many strings, arrays, etc. are allocated.

Envek commented:
@pixeltrix I do not know how to see object count, can you please post some link to guide or anything?

pixeltrix commented:
Here's some tools you can use:

https://github.com/srawlins/allocation_stats
https://github.com/ko1/allocation_tracer
https://github.com/SamSaffron/memory_profiler

Or you can simply compare `ObjectSpace.count_objects` before and after the call to get a rough idea:
http://ruby-doc.org/core-2.2.3/ObjectSpace.html#method-c-count_objects

Envek committed:
Switched from RegExp based parser to StringScanner based one

Envek commented:
@pixeltrix I changed parser to your `StringScanner` based one. Also added tests for cases when individual parts have different signs. Please take a look.

I executed against 2f189b0 next snippet against both regexp and strscan implementations:
```ruby
require_relative 'lib/active_support/all'
require 'strscan'

before = nil
after = nil
diff = {}
before = ObjectSpace.count_objects
ActiveSupport::Duration.parse('P1Y2M3DT4H5M6S')
after = ObjectSpace.count_objects

before.keys.each do |k|
  diff[k] = after[k] - before[k]
end

puts diff.to_yaml
```

And got next results:

| key        | strscan | regexp |
| ---------- | ------: | ------:|
| TOTAL      |       0 |      0 |       
| FREE       |  -25948 | -25626 |           
| T_OBJECT   |     217 |    217 |            
| T_CLASS    |      96 |     95 |          
| T_MODULE   |      14 |     14 |           
| T_FLOAT    |       0 |      0 |         
| T_STRING   |   10456 |  10341 |              
| T_REGEXP   |      56 |     52 |           
| T_ARRAY    |    2085 |   2073 |            
| T_HASH     |     157 |    155 |          
| T_STRUCT   |       1 |      2 |          
| T_BIGNUM   |       0 |      0 |          
| T_FILE     |      45 |     45 |         
| T_DATA     |     734 |    727 |          
| T_MATCH    |     147 |    148 |           
| T_COMPLEX  |       0 |      0 |           
| T_RATIONAL |       1 |      1 |            
| T_NODE     |   11904 |  11721 |            
| T_ICLASS   |      14 |     14 |           

Seems like difference between approaches is really negligible.

pixeltrix commented:
@Envek are those numbers the difference, i.e. after - before for each implementation. If so they seem odd - for example I wouldn't thought it'd create 45 new file objects. :confused:

pixeltrix commented on activesupport/lib/active_support/duration/iso8601_parser.rb:
```diff
@@ -0,0 +1,138 @@
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code and logic are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601Parser
```
Probably want to `:nodoc:` this class and any other implementation details.

pixeltrix commented on activesupport/lib/active_support/duration/iso8601_parser.rb:
```diff
@@ -0,0 +1,138 @@
+            nil
+          end
+        end
+      end
```
We don't use these class methods do we so we might as well drop them

Envek committed:
Removed unused class methods, added :nodoc:

Envek commented:
@pixeltrix removed unused class methods, added :nodoc:
Any other issues?

Envek commented:
Yes, those numbers are the difference. And yes they looks weird.

pixeltrix commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
+    def self.parse!(iso8601duration)
+      parts = ISO8601Parser.new(iso8601duration).parse!
+      time  = ::Time.current
+      new(time.advance(parts) - time, parts)
```
What happens when we go across a DST boundary here? Does it affect the value passed in to `new`?

Envek replied:
Yes, it affects.

```
2.2.3 :002 > parts = ActiveSupport::Duration::ISO8601Parser.new('-P1Y3M').parse!
 => {:years=>-1, :months=>-3}
2.2.3 :003 > time_local = Time.current
 => 2015-11-04 18:01:53 +0300
2.2.3 :004 > time_utc   = time_local.dup.utc
 => 2015-11-04 15:01:53 UTC
2.2.3 :005 > time_local.advance(parts)
 => 2014-08-04 18:01:53 +0400
2.2.3 :006 > time_utc.advance(parts)
 => 2014-08-04 15:01:53 UTC
2.2.3 :007 > duration_local = ActiveSupport::Duration.new(time_local.advance(parts) - time_local, parts)
 => -1 years and -3 months
2.2.3 :008 > duration_local.value
 => -39488400.0
2.2.3 :009 > duration_utc = ActiveSupport::Duration.new(time_utc.advance(parts) - time_utc, parts)
 => -1 years and -3 months
2.2.3 :010 > duration_utc.value
 => -39484800.0
2.2.3 :011 > time_local.advance(parts) - time_local
 => -39488400.0
2.2.3 :012 > time_utc.advance(parts) - time_utc
 => -39484800.0
2.2.3 :013 > (-1.year - 3.months).value
 => -39333600.0
2.2.3 :015 > time_local + duration_local
 => 2014-08-04 18:01:53 +0400
2.2.3 :016 > time_utc + duration_utc
 => 2014-08-04 15:01:53 UTC
2.2.3 :017 > time_utc + duration_local
 => 2014-08-04 15:01:53 UTC
2.2.3 :018 > time_local + duration_utc
 => 2014-08-04 18:01:53 +0400
```

1 year and 3 months ago in Moscow was summer time in action. So I created both utc and local times, created durations, examined them and done some math with these duration. Despite the fact, that numeric value of seconds passed to `new` in each case is different (and both are different from “hand made” duration), math results are the same.

I think it's better to stick with `Time.current`, because it's better to have information about local time offset than not to have. If someone want to explicitly work with UTC he should set `Time.zone = 'UTC'` in initializer.

Listing of this test:
```ruby
require_relative 'lib/active_support/all'
Time.zone = 'Moscow'
parts = ActiveSupport::Duration::ISO8601Parser.new('-P1Y3M').parse!
time_local = Time.current
time_utc   = time_local.dup.utc
time_local.advance(parts)
time_utc.advance(parts)
duration_local = ActiveSupport::Duration.new(time_local.advance(parts) - time_local, parts)
duration_local.value
duration_utc = ActiveSupport::Duration.new(time_utc.advance(parts) - time_utc, parts)
duration_utc.value
time_local.advance(parts) - time_local
time_utc.advance(parts) - time_utc
(-1.year - 3.months).value
time_local + duration_local
time_utc + duration_utc
time_utc + duration_local
time_local + duration_utc
```

Envek replied:
@pixeltrix What do you think? I can't figure out which variant is correct. For me `Time.current` is sane.

Or may be it's better not to use `Time.advance` at all but manually calculates value by doing something like this:

```ruby
parts = ActiveSupport::Duration::ISO8601Parser.new('P1Y2M3DT4H5M6S').parse!
duration = parts.map{ |unit, value| value.send(unit) } # 1.year + 2.months + 3.days + …
duration.inspect # 1 year, 2 months, 3 days, and 14706 seconds
duration.value   # 37015506.0 — different from current parse method, but stable.
duration.iso8601 # "P1Y2M3DT14706S" — hours and minutes are lost
```

But in such scenario there are lot of intermediate durations created meanwhile (as many as parts in duration, 6 in this case, not good).

sgrif commented:
r? @pixeltrix Are you good with this being merged? PG interval support requires this, and I'd like to get that in for Rails 5

pixeltrix commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -220,4 +220,71 @@ def test_comparable
+  end
+
+  def test_iso8601_output
+    # As (1.year + 1.month + 1.day + 1.hour).iso8601 will output P1Y1M1DT3600S (weird), test for these not included
```
I think even if they're weird we should still test them to make sure that the values don't change unexpectedly

pixeltrix commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -220,4 +220,71 @@ def test_comparable
+    # As (1.year + 1.month + 1.day + 1.hour).iso8601 will output P1Y1M1DT3600S (weird), test for these not included
+    expectations = [
+      ['P1Y',        1.year                    ],
+      #['P1W',        1.week                    ], # Actually 1.week returns 7 days
```
I think we should still include the output test for 1 week - just the expected value should be `P7D`

pixeltrix commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -220,4 +220,71 @@ def test_comparable
+      assert_equal time+duration, time+ActiveSupport::Duration.parse(duration.iso8601), pattern.inspect
+    end
+  end
+
```
Cut this line space, thanks

pixeltrix commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -80,7 +80,7 @@ def ago(time = ::Time.current)
       parts.
         reduce(::Hash.new(0)) { |h,(l,r)| h[l] += r; h }.
-        sort_by {|unit,  _ | [:years, :months, :days, :minutes, :seconds].index(unit)}.
+        sort_by {|unit,  _ | [:years, :months, :weeks, :days, :hours, :minutes, :seconds].index(unit)}.
```
My only concern is that we're going to get different output depending on how the duration is constructed, e.g:

``` irb
>> ActiveSupport::Duration.parse('P1W')
=> 1 week
>> 1.week
=> 7 days
```

Pondering whether we should not decompose values in Rails 5.

Envek replied:
I can try to fix that in some another pull request in a few days.

Envek committed:
Add tests for weird durations, cut the line

Envek committed:
Oh no, Time.now again!

Envek commented:
@pixeltrix I fixed your today notes. If there is nothing else left I will be glad to squash and rebase. Thanks for review.

nambrot commented:
Just wanted to give another thumbs up for this one. I'll be implementing this external, but it would have been very nice to be part of ActiveSupport

Envek commented:
@nambrot thank you for your support. 

You can just use serializer from my gist that will enable ISO8601 parsing and output: https://gist.github.com/Envek/7077bfc36b17233f60ad

Meanwhile I'm still wating for feedback in this PR as I don't know whether I should change something.

jeremy commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
+    # https://google-styleguide.googlecode.com/svn/trunk/jsoncstyleguide.xml?showone=Time_Duration_Property_Values#Time_Duration_Property_Values
     def as_json(options = nil) #:nodoc:
-      to_i
+      iso8601(precision: options && options[:precision])
```
I'm afraid this will significantly break app compatibility.

We'll need a config flag for Duration JSON representation that's set the iso8601 for new apps but defaults to integer seconds for existing apps.

(Check git history for how we've tackled this before)

jeremy replied:
Actually, it'd be far simpler to just omit this. Leave the JSON representation as-is and let people opt in to calling `duration.iso8601` if they wish. That'd significantly lower the bar (and additional work) for merging the PR.

Envek replied:
Reverted JSON representation to initial `to_i`

jeremy commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601Parser::ParsingError+.
+    def self.parse!(iso8601duration)
```
`parse!` → `parse`

This needn't be a bang method.

jeremy commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
+      parse!(iso8601duration)
+    rescue ISO8601Parser::ParsingError
+      nil
+    end
```
IMO we don't need a method that swallows the parse error.

jeremy commented:
If we're pulling in code under MIT license, we need to include the license and copyright statement.

Also, if we're pulling in code, perhaps we should do a gem dep instead?

Nice work on this, and patience :grin:

Envek commented:
@jeremy there is not a much code left from the original code from ISO8601 gem as I've rewritten parser from regexp to string scanner by @pixeltrix suggestion. As far as I can see for now only duration examples strings inside tests are same with ISO8601. See the original [code](https://github.com/arnau/ISO8601/blob/master/lib/iso8601/duration.rb) and [tests](https://github.com/arnau/ISO8601/blob/master/spec/iso8601/duration_spec.rb).

Anyway where I should place license and copyright statement if I should? @arnau what do you think?

Gem dependency means that we should throw away whole `ActiveSupport::Duration` and replace it with `ISO8601::Duration`. Are you serious with that? :smile: 

@jeremy please take a look at https://github.com/rails/rails/pull/22806 while I'm fixing your notes— it's extracted from this pull request.

Envek committed:
Revert "Changed duration output as json to ISO 8601 duration format"

This reverts commit c130b5c35aed30b8b8b4e6744448690856f262e5.

Envek committed:
Removed error-swallowing parsing method from duration. parse! → parse

Envek commented:
@jeremy fixed your notes

jeremy commented:
@Envek We can't replace AS::Duration with ISO8601::Duration, but perhaps we could delegate implementation to it. You tell me :)

If you've rewritten the code, then you're no longer using the original code and aren't redistributing it, so you needn't pull in its license.

If you are reusing code, include the license+copyright along with the code, in a Ruby comment.

arnau commented:
@Envek do whatever makes more sense to you :)

Envek committed:
Merge branch 'master' into iso8601_duration

Envek committed:
Change notices about licensing status and code and data usage to reflect current status

Envek commented:
@jeremy, @arnau, I changed notices about licensing, take a look at https://github.com/rails/rails/pull/16917/commits/3a92929d83bffce256b64ebb59354ad04a11fc4f . Is it enough?

If yes I would be glad to rebase and squash.

jeremy commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -224,6 +224,9 @@ def test_comparable
 
+  # ISO8601 string examples are taken from ISO8601 gem at https://github.com/arnau/ISO8601/blob/b93d466840/spec/iso8601/duration_spec.rb
+  # published under the conditions of MIT license: https://github.com/arnau/ISO8601/blob/b93d466840/LICENSE
+
```
The license is pretty clear about what we need to include here:
```
The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.
```
So we need to include the license text in a comment here.

Envek replied:
Included the license text in a comment

Envek committed:
Include the whole MIT license to comments for the duration tests.

vipulnsward commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -130,6 +133,23 @@ def respond_to_missing?(method, include_private=false) #:nodoc:
 
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
```
`See http://en.wikipedia.org/wiki/ISO_8601#Durations .`

vipulnsward commented on activesupport/lib/active_support/duration/iso8601_parser.rb:
```diff
@@ -0,0 +1,122 @@
+  class Duration
+    # Parses a string formatted according to ISO 8601 Duration into the hash .
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
```
`http://en.wikipedia.org/wiki/ISO_8601#Durations.`

Envek replied:
Placing dot in end of URL will make it broken.

vipulnsward replied:
Actually this should be wrapped in as a markdown link.
`See [ISO 8601](http://en.wikipedia.org/wiki/ISO_8601#Durations) for more information.`

Envek replied:
Links have been markdownified.

vipulnsward commented on activesupport/lib/active_support/duration/iso8601_serializer.rb:
```diff
@@ -0,0 +1,51 @@
+
+      private
+
+        # Return pair of duration's parts and whole duration sign
```
missing `.`. There are a couple of places its missing, @Envek can you take a look?

Envek replied:
Fixed!

vipulnsward commented on activesupport/lib/active_support/duration/iso8601_serializer.rb:
```diff
@@ -0,0 +1,51 @@
+
+        # Return pair of duration's parts and whole duration sign
+        # Parts are summarized (as they can become repetitive due to addition, etc)
+        # zero parts are removed as not significant.
```
`Zero`

Envek committed:
Dots. Dots are everywhere.

Envek committed:
Markdonw links in comments to Wikipedia article

vipulnsward commented on activesupport/lib/active_support/duration/iso8601_parser.rb:
```diff
@@ -0,0 +1,122 @@
+  class Duration
+    # Parses a string formatted according to ISO 8601 Duration into the hash.
+    #
+    # See [ISO 8601](http://en.wikipedia.org/wiki/ISO_8601#Durations) for more information.
```
😢 Sorry this is rdoc, my bad. Let me see.

vipulnsward replied:
So rdoc makes sure the link is proper even if it's followed by a '.' Anyway, lets change to -

`See {ISO 8601}[http://en.wikipedia.org/wiki/ISO_8601#Durations] for more information.`

vipulnsward commented on activesupport/lib/active_support/duration.rb:
```diff
@@ -130,6 +133,23 @@ def respond_to_missing?(method, include_private=false) #:nodoc:
 
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See [ISO 8601](http://en.wikipedia.org/wiki/ISO_8601#Durations) for more information.
```
`See {ISO 8601}[http://en.wikipedia.org/wiki/ISO_8601#Durations] for more information.`

Envek replied:
RDocified links, thanks.

Envek committed:
Markdown links → RDoc links

jeremy commented:
Merged! https://github.com/rails/rails/commit/04c512da1247a54474cfd8bef17a9e9019c34004

Envek commented:
@jeremy thank you for merging! You have just unlocked a special #16919!

P.S> I have a one more small ActiveSupport goodness: #20625, you can review it too if you have free time and wish :-)

Dorian commented:
This makes using duration in `where` condition easy, thanks @Envek !!!

```
House.where("(houses.sold_at - searches.bought_at) >= ?", 20.years.iso8601).size
```
